### PR TITLE
reorganize mesh and material hierarchies, fix lab 7

### PR DESCRIPTION
**Pull request by JoseRivas1998**

### CODE REVIEW

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    public abstract void draw();
+    protected abstract void activate();
+    protected abstract void deactivate();
+
+    protected AbstractMesh(Vector3 position, Quaternion rotation, Vector3 scale) {
+        super(position, rotation, scale);
+    }
+
+    protected void activateAttribute3f(AbstractAttribute attribute, int attributeLocation) {
```
Comment (JoseRivas1998) [inline src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java:31]: These functions seem to only be used by unused functions, can you explain their purpose?

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AttributeColorMesh.java [hunk 2]
```diff
@@ -67,9 +57,17 @@ private void createColorBuffer() {
...
+        this.activateAttribute4f(attribute, attributeLocation);
+    }
+
     @Override
     public void dispose() {
         glDeleteBuffers(this.colorVboId);
         super.dispose();
     }
+
+    public static enum Attribute implements AbstractAttribute {
```
Comment (JoseRivas1998) [inline src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AttributeColorMesh.java]: Declaring an inner enum as static is redundant

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/Mesh.java [hunk 3]
```diff
@@ -64,8 +64,10 @@ private void createBuffers() {
             positionBuffer = MemoryUtil.memAllocFloat(this.positionArray.length);
             positionBuffer.put(this.positionArray).flip();
 
-            this.positionAttribLocation = this.material.shaderProgram.getAttribLocation("vertPosition");
-            this.positionVboId = glGenBuffers();
+            this.attributeBufferMap = new HashMap<>();
+
+            this.positionAttribLocation = this.shaderProgram.getAttribLocation("vertPosition");
+            this.positionVboId = createAttributeBuffer(Attribute.POSITION);
```
Comment (JoseRivas1998) [inline src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/Mesh.java]: Please use `this`. notation.
You can enable IntelliJ to warn when `this.` is not used by enabling the following settings:

File > Settings > [Project Settings] > Inspections > Code Style Issues >
- "Instance method call not qualified with 'this'"
- "Instance fields access not qualified with 'this'"

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import com.tcg.lwjgllearning.math.Transform3D;
+import com.tcg.lwjgllearning.math.Vector3;
+import com.tcg.lwjgllearning.utils.Disposable;
+
+import static org.lwjgl.opengl.GL15.*;
+import static org.lwjgl.opengl.GL20.*;
+
+public abstract class AbstractMesh extends Transform3D implements Disposable, Drawable {
+    
+    protected HashMap<AbstractAttribute, Integer> attributeBufferMap;
```
Comment (JoseRivas1998) [inline src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java:16]: I can see where this is populated, but where/how is it used?

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java (hunk 1 above)
Comment (arewhyaeenn) [reply src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java:31]: The unused functions that use those are intended to be used when you need particular attribute(s) to be used in a draw call with a shader and/or vao other than those associated with the mesh containing said attributes. For example, when creating a particular light's depth buffer for shadow mapping, a very simple shader and a vao other than the mesh's will be used, but the mesh's position attribute will be needed.

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AttributeColorMesh.java (hunk 2 above)
Comment (arewhyaeenn) [reply src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AttributeColorMesh.java]: Oh, good to know, didn't realize! Will remove the redundancy.

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/Mesh.java (hunk 3 above)
Comment (arewhyaeenn) [reply src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/Mesh.java]: Oops, missed one! I'm using VSCode, but I will find the setting there ^^

File: src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java (hunk 4 above)
Comment (arewhyaeenn) [reply src/main/java/com/tcg/lwjgllearning/graphics/g3d/mesh/AbstractMesh.java:16]: It is used in the "unused" methods discussed in your first comment above. Its purpose is to allow other classes to specify an attribute in said methods without having access to the vbo identifier.

**Pull request merged**